source/qnx/ionotify_event_loop.c

```diff
@@ -0,0 +1,1042 @@
+    aws_mutex_unlock(&ionotify_loop->task_pre_queue_mutex);
+
+    /* If the list was not empty, we already sent a cross-thread pulse. No need to send it again. */
+    if (is_first_task) {
+        AWS_LOGF_TRACE(
+            AWS_LS_IO_EVENT_LOOP,
+            "id=%p: Waking up event-loop thread by sending pulse to connection ID %d",
+            (void *)event_loop,
+            ionotify_loop->pulse_connection_id);
+        /* The pulse itself is enough for cross-thread notifications. */
+        int user_data_value = 0;
+        int rc = MsgSendPulse(ionotify_loop->pulse_connection_id, -1, CROSS_THREAD_PULSE_SIGEV_CODE, user_data_value);
+        int errno_value = errno;
+        if (rc == -1) {
+            /* The task was scheduled, but we couldn't notify the main loop about it. According to QNX docs, inability
+             * to send a pulse indicates that there is no available memory left for the process. Not notifying the loop
+             * is the minor thing in such a scenario. So, just log the error. */
+            AWS_LOGF_ERROR(
+                AWS_LS_IO_EVENT_LOOP,
+                "id=%p: Failed to send cross-thread pulse: %d (%s)",
+                (void *)event_loop,
+                errno_value,
+                strerror(errno_value));
+        }
+    }
+}
+
+static void s_schedule_task_now(struct aws_event_loop *event_loop, struct aws_task *task) {
+    s_schedule_task_common(event_loop, task, 0 /* zero denotes "now" task */);
+}
+
+static void s_schedule_task_future(struct aws_event_loop *event_loop, struct aws_task *task, uint64_t run_at_nanos) {
+    s_schedule_task_common(event_loop, task, run_at_nanos);
+}
+
+static void s_cancel_task(struct aws_event_loop *event_loop, struct aws_task *task) {
+    AWS_LOGF_TRACE(AWS_LS_IO_EVENT_LOOP, "id=%p: Cancelling task %p", (void *)event_loop, (void *)task);
+    struct ionotify_loop *ionotify_loop = event_loop->impl_data;
+    aws_task_scheduler_cancel_task(&ionotify_loop->scheduler, task);
+}
+
+/* Map ionotify_event_data to internal ID. */
+static int s_add_handle(struct ionotify_loop *ionotify_loop, struct ionotify_event_data *ionotify_event_data) {
+    AWS_ASSERT(s_is_on_callers_thread(ionotify_event_data->event_loop));
+
+    /* Special constant, _NOTIFY_DATA_MASK, limits the maximum value that can be used as user data in I/O events. */
+    int max_handle_id = _NOTIFY_DATA_MASK;
+
+    if (AWS_UNLIKELY(aws_hash_table_get_entry_count(&ionotify_loop->handles) == (size_t)max_handle_id)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_EVENT_LOOP,
+            "id=%p: Maximum number of registered handles reached",
+            (void *)ionotify_event_data->event_loop);
+        return AWS_OP_ERR;
+    }
+
+    struct aws_hash_element *elem = NULL;
+    int next_handle_id = ionotify_loop->last_handle_id;
+    int was_created = 0;
+    do {
+        ++next_handle_id;
+        if (next_handle_id > max_handle_id) {
+            next_handle_id = 1;
```

Comment:
Yes, I decided not to use 0 as a valid `handle_id` because
1. This handle is passed with the events from QNX resource managers, and I want (probably, irrationally) to have at least 1 bit set. In my head, it helps catch a case where a resource manager forgets to return user-provided data.
2. `s_subscribe_task()` is used both for subscribing and resubscribing to the I/O events. `0` helps with distinguishing these two cases. But this point was just a follow-up on the already made decision.